### PR TITLE
Feat: Add Deref Coercion

### LINKED ISSUE #890: feat: Deref coercion
The mechanism is similar to Rust's, but the scope is a lot smaller: trivially allow access to a member of T holding a type U.
https://doc.rust-lang.org/book/ch15-02-deref.html#treating-a-type-like-a-reference-by-implementing-the-deref-trait

Should go in 11. Advanced Features
Demonstration here: compile and read the casm to see what's happening
```
#[derive(Copy, Drop)]
struct MyStruct {
    x: felt252,
    y: felt252,
    z: felt252,
}

impl DerefType1ToType2 of Deref<Box<MyStruct>> {
    type Target = MyStruct;
    fn deref(self: Box<MyStruct>) -> Self::Target {
        let unboxed = self.unbox();
        Self::Target { x: unboxed.x, y: unboxed.y, z: unboxed.z }
    }
}


fn main() {
    let my_struct = MyStruct { x: 1, y: 2, z: 3 };
    let boxed = BoxTrait::new(my_struct);

    foo(boxed);
    bar(boxed);
}

#[inline(never)]
fn foo(boxed: Box<MyStruct>) -> felt252 {
    boxed.x
}

#[inline(never)]
fn bar(boxed: Box<MyStruct>) -> felt252 {
    boxed.unbox().x
}
```

### PR DESCRIPTION
**Pull request by blocksorcerer**
Closes #890

- - -
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/897)

### CODE REVIEW

**Comment by enitrat** [comment]
Lots of CI errors due to `no_listing_09_deref_coercion_example`, please check

File: src/ch11-09-deref-coercion.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Deref Coercion
+
+In Cairo, deref coercion simplifies the way we interact with nested or wrapped data structures by allowing an instance of one type to behave like an instance of another type.   
+This mechanism is enabled by implementing the `Deref` trait, which allows implicit conversion (or coercion) of references to a different type, providing direct access to the underlying data. Additionally, the `DerefMut` trait allows mutable references to be coerced similarly, Enabling direct modification of the target data.
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:4]: Careful about terminology: How could you rephrase this to avoid using the term `mutable reference` (and more generally `reference`) ?

File: src/ch11-09-deref-coercion.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+## Understanding Deref Coercion
+
+The `Deref` and `DerefMut` traits are used to override the behavior of the dereference operator `*`. When implemented for a type, they allow an instance of that type to automatically convert to its target type, as specified by the traits. This makes accessing fields or calling methods on the target type seamless and intuitive.
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:8]: This is factually not true: Deref and DerefMut have no impact on the `*` operator (at least, yet)

File: src/ch11-09-deref-coercion.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+### The DerefMut Trait
+
+```rust, noplayground
+{{#rustdoc_include ../listings/ch11-advanced-features/listing_09_deref_coercion/src/lib.cairo:DerefMut}}
+```
+
+Similarly, the `DerefMut` trait is used for mutable dereferencing. When you dereference a mutable value of type `T`, the compiler looks for an implementation of `DerefMut` for `T`. If found, it applies the `deref_mut` method to obtain a mutable reference to the target type.
+
+## Deref and DerefMut Coercion in Action
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:25]: The example provided is a good start, but it can be further improved to display a stronger use case combined with generic types.

For example, this could be defining our own "wrapper type" like the Box<T> type, that would simply wrap any data of type T `MyBox<T>`.

By implementing a generic implementation of the Deref<MyBox<T>> type, we can demonstrate how we can easily abstract the complexity of any generic wrapper type by implementing this Deref Trait - and that's mostly how it's being used in the core library (e.g. for `Nullable`, `Box`, `StorageNode`, etc...)

for reference, this is how it's presented in the Rust Book: https://doc.rust-lang.org/book/ch15-02-deref.html?highlight=deref#defining-our-own-smart-pointer

File: src/ch11-09-deref-coercion.md (hunk 1 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md:4]: How about something like following:-

In Cairo, deref coercion simplifies the way we interact with nested or wrapped data structures by allowing an instance of one type to behave like an instance of another type. This mechanism is enabled by implementing the `Deref` trait, which allows implicit conversion (or coercion) to a different type, providing direct access to the underlying data. Additionally, the `DerefMut` trait allows similar coercion, enabling direct modification of the target data. For now, only member access via deref is supported, i.e. implementations with functions whose self argument is of the original type will not be applicable when holding an instance of the coerced type.

File: src/ch11-09-deref-coercion.md (hunk 2 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md:8]: Thanks for letting me know, I've changed this.

File: src/ch11-09-deref-coercion.md (hunk 3 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md:25]: Can't access this `DerefMut` trait. Getting trait not exists error! 
Also I've changed the example to show a generic wrapper type.

File: src/ch11-09-deref-coercion.md (hunk 1 above)
Comment (enitrat) [reply src/ch11-09-deref-coercion.md:4]: looks good

File: src/ch11-09-deref-coercion.md (hunk 3 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md:25]: @enitrat any suggestions on this one?? Getting `error: Trait not found.` for DerefMut

Comment (enitrat) [reply src/ch11-09-deref-coercion.md:25]: did you import it ?

Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md:25]: I haven't... Import needed? Deref is working fine that's why!

**Comment by enitrat** [comment]
so actually DerefMut is not like Rust's - its usage is only for cases where you want deref to work only for a mut variable

File: src/ch11-09-deref-coercion.md (hunk 2 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md]: Please review d76aaf7

File: src/ch11-09-deref-coercion.md (hunk 1 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md]: Fixed in d76aaf7

File: listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#[inline(never)]
+fn update_age(ref wrapped_profile: Wrapper<UserProfile>, age: u16) {
+    // Mutably access the UserProfile via deref_mut coercion
+    let updated_profile = Wrapper {
+        value: UserProfile {
+            username: wrapped_profile.username, email: wrapped_profile.email, age: age
+        }
+    };
+    wrapped_profile = updated_profile;
+}
```
Comment (enitrat) [inline listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo:85]: remove all these functions, you can simply inline stuff in main

File: listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    // Access and print the age directly
+    print_age(wrapped_profile);
+
+    // Update the age
+    update_age(ref wrapped_profile, 25);
+
+    // Print the updated age
+    print_age(wrapped_profile);
+}
```
Comment (enitrat) [inline listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo:60]:
```suggestion
fn main() {
    let wrapped_profile = Wrapper { 
        value: UserProfile { 
            username: 'john_doe', 
            email: 'john@example.com', 
            age: 30 
        } 
    };

    // Access fields directly via deref coercion
    println!("Username: {}", wrapped_profile.username);
    println!("Current age: {}", wrapped_profile.age);
}
```

File: src/ch11-09-deref-coercion.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+In Cairo, deref coercion simplifies the way we interact with nested or wrapped data structures by allowing an instance of one type to behave like an instance of another type. This mechanism is enabled by implementing the `Deref` trait, which allows implicit conversion (or coercion) to a different type, providing direct access to the underlying data. Additionally, the `DerefMut` trait allows similar coercion, enabling direct modification of the target data. For now, only member access via deref is supported, i.e. implementations with functions whose `self` argument is of the original type will not be applicable when holding an instance of the coerced type.
+
+## Understanding Deref Coercion
+
+In Cairo, the `Deref` and `DerefMut` traits allow you to customize the behavior of accessing members of a type through another type. When a type `T` implements `Deref` or `DerefMut` to type `K`, instances of `T` can access the members of `K` directly.  
+
+Let's look at the `Deref` and `DerefMut` traits closely :-
+
+### The Deref Trait
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:11]:
```suggestion
### Deref in Action
```

Move the example of using `Deref` (and not DerefMut) here

File: src/ch11-09-deref-coercion.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+The `Deref` trait in Cairo is defined as follows:
+
+```rust, noplayground
+{{#rustdoc_include ../listings/ch11-advanced-features/listing_09_deref_coercion/src/lib.cairo:Deref}}
+```
+Where:
+- `type Target`: Specifies the target type `K` that the type `T` will dereference to.
+- `fn deref(self: T) -> Self::Target`: This method allows you to define how an instance of type `T` converts into type `K`. When implemented, you can access members of `K` through an instance of `T`.
+
+### The DerefMut Trait
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:22]:
```suggestion
### Restricting Deref Coercion with `DerefMut`
```

File: src/ch11-09-deref-coercion.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+The `DerefMut` trait works similarly but is used for mutable access:
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:24]: update with the proper definition as per my previous comment

File: src/ch11-09-deref-coercion.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+```rust, noplayground
+{{#rustdoc_include ../listings/ch11-advanced-features/listing_09_deref_coercion/src/lib.cairo:DerefMut}}
+```
+Where:
+- `fn deref_mut(ref self: T) -> Self::Target`: This method allows mutable access to the target type `K` through a mutable reference to type `T`.
+
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:31]: Here you must explicitly illustrate what's the difference between Deref and DerefMut. `Deref` will work for non-mut AND mut variables, while DerefMut will only work with non-mut variables.

In practice DerefMut probably doesn't have a lot of use cases. Thus you should illustrate that if you try to  use deref-coercion, having implemented DerefMut, on a non-mut variable, compilation fails, see following example

```rust
impl DerefMutUserProfile of core::ops::DerefMut<Wrapper<UserProfile>> {
    type Target = UserProfile;
    fn deref_mut(ref self: Wrapper<UserProfile>) -> Self::Target {
        self.value
    }
}

fn main() {
    let wrapped_profile = Wrapper {
        value: UserProfile { username: 'john_doe', email: 'john@example.com', age: 30 }
    };
    // Access fields directly via deref coercion
    println!("Username: {}", wrapped_profile.username);
    println!("Current age: {}", wrapped_profile.age);
}
```

File: src/ch11-09-deref-coercion.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+{{#rustdoc_include ../listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo:deref}}
+```
+With these implementations, we can now access and modify the value inside `Wrapper<T>` directly, without needing to manually unwrap it.  
+
+Let's see how this works in practice by using `Wrapper` to wrap a `UserProfile` and demonstrate how `Deref` and `DerefMut` make it easier to work with the wrapped data.
+
+```rust, noplayground
+{{#rustdoc_include ../listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo:main}}
+```
+By using the `Deref` and `DerefMut` traits, we can create a generic wrapper type like `Wrapper<T>` that allows seamless access to the underlying data. This pattern is common in core libraries for abstracting away the complexity of working with various data types while maintaining clear and concise code.
```
Comment (enitrat) [inline src/ch11-09-deref-coercion.md:60]: See comments above regarding re-structuration of this example

**Review by enitrat — CHANGES_REQUESTED** [review]
It's getting better!

Some restructuring still required, and all the part about `DerefMut` does not explain what it actually does.

File: src/ch11-09-deref-coercion.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+```rust, noplayground
+{{#rustdoc_include ../listings/ch11-advanced-features/listing_09_deref_coercion/src/lib.cairo:DerefMut}}
+```
+Where:
+- `fn deref_mut(ref self: T) -> Self::Target`: This method allows mutable access to the target type `K` through a mutable reference to type `T`.
+
+## Deref and DerefMut Coercion in Action
+
+Let’s consider a generic wrapper type, similar to a smart pointer, that can wrap any data of type `T`. By implementing the `Deref` and `DerefMut` traits generically, we can abstract away the complexity of accessing the inner data, much like how it’s done in core libraries for types like `Box`, `Nullable`, and `StorageNode`.
```
[the remark this answers is not in the document]
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md]: changed in b73a359

File: src/ch11-09-deref-coercion.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+{{#rustdoc_include ../listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo:UserProfile}}
+```
+The `UserProfile` struct represents a user profile with fields `username`, `email`, and `age`.  
+
+Let's define a generic `Wrapper<T>` type and demonstrate how `Deref` and `DerefMut` can be used to access the wrapped data effortlessly.
+
+```rust, noplayground
+{{#rustdoc_include ../listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo:Wrapper}}
+```
+The `Wrapper` struct wraps a single value of type `T`. Normally, accessing this value would require going through `Wrapper.value`.
```
[the remark this answers is not in the document]
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md]: changed in b73a359

File: listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo (hunk 4 above)
Comment (blocksorcerer) [reply listings/ch11-advanced-features/no_listing_09_deref_coercion_example/src/lib.cairo]: Implemented in b73a359

File: src/ch11-09-deref-coercion.md (hunk 9 above)
Comment (blocksorcerer) [reply src/ch11-09-deref-coercion.md:27]: Please check b73a359

**Review by enitrat — DISMISSED** [review]
